tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Plan.java

```diff
@@ -56,13 +56,14 @@
   long timeout;
 
   @Override
-  public void execute() throws MojoExecutionException, MojoFailureException {
+  public void execute() throws MojoExecutionException {
     try {
       if (!StringUtils.isEmpty(artifact)) {
         TerraformGetMavenRootArtifact mavenRepoExecutableOp = new TerraformGetMavenRootArtifact(artifact, tfRootDir, getLog());
         tfRootDir = mavenRepoExecutableOp.execute(getFieldsAsProperties());
       }
-      execute(new TerraformPlan(tfRootDir, new MavenSlf4jAdapter(getLog())));
+      execute(new TerraformPlan(tfRootDir, new MavenSlf4jAdapter(getLog())), System.getProperties());
```

Comment:
Why not something like this?
```java
boolean isS3Upload = planOutputFile..toUpperCase().startsWith("S3://");
String planOutputFilename = isS3Upload ? planOutputFile.substring(planOutputFile.lastIndexOf("/")) : planOutputFile;
Properties props = new Properties();
props.put("planOutputFile", planOutputFilename)
execute(new TerraformPlan(tfRootDir, new MavenSlf4jAdapter(getLog())), getFieldsAsProperties(props));
if (isS3Upload) {
   execute(new TerraformS3Upload(tfRootDir, new MavenSlf4jAdapter(getLog())), getFieldsAsProperties());
}

```

That way, you can keep your changes localized to the Mojo (UI) logic. Volatility should be localized - it's always easier to require changes in one place than to require them in multiple places.

Of course to do that, you will have to overload the getFieldsAsProperties() method in the TerraformMojo class.

```java
  protected final Properties getFieldsAsProperties(Properties props) throws MojoExecutionException {
    Properties properties = getFieldsAsProperties();
    properties.putAll(props);
    return properties;
  }
```